doc/manual/manual.md

```diff
@@ -2476,3 +2476,249 @@ Currently, this function can only be used for LLVM verification, though
+~~~~
+
+#### 32-Bit Key Expansion
+
+The next function of substantial behavior that we wish to verify has type
+
+~~~~c
+void s20_expand32( uint8_t *k
+                 , uint8_t n[static 16]
+                 , uint8_t keystream[static 64]
+                 )
+~~~~
+
+This function's specification follows a similar pattern to that of
+`s20_quarterround`, though for extra assurance we can make sure that the
+function does not write to the memory pointed to by `k` or `n` using the
+utility `ptr_to_fresh_readonly`, as this function should only modify
+`keystream`. Besides this, we see the call to the trusted Cryptol
+implementation specialized to `a=2`, which does 32-bit key expansion (since the
+Cryptol implementation can also specialize to `a=1` for 16-bit keys.) This
+specification can easily be changed to work with 16-bit keys.
+
+~~~~
+let salsa20_expansion_32 = do {
+    (k, pk) <- ptr_to_fresh_readonly "k" (llvm_array 32 (llvm_int 8));
+    (n, pn) <- ptr_to_fresh_readonly "n" (llvm_array 16 (llvm_int 8));
+
+    pks <- crucible_alloc (llvm_array 64 (llvm_int 8));
+
+    crucible_execute_fun [pk, pn, pks];
+
+    let rks = {{ Salsa20_expansion`{a=2}(k, n) }};
+    crucible_points_to pks (crucible_term rks);
+};
+~~~~
+
+#### 32-bit Key Encryption
+
+Finally, we write a specification for the encryption function itself, which has
+type
+
+~~~~c
+enum s20_status_t s20_crypt32( uint8_t *key
+                             , uint8_t nonce[static 8]
+                             , uint32_t si
+                             , uint8_t *buf
+                             , uint32_t buflen
+                             )
+~~~~
+
+As before, we can ensure this function does not modify the memory pointed to by
+`key` or `nonce`. We take `si`, the stream index, to be 0. The specification is
+parameterized on a number `n`, which corresponds to `buflen`. Finally, to deal
+with the fact that this function returns a status code, we simply specify that
+we expect a success (status code 0) as the return value in the postcondition
+stage of the specification.
+
+~~~~
+let s20_encrypt32 n = do {
+    (key, pkey) <- ptr_to_fresh_readonly "key" (llvm_array 32 (llvm_int 8));
+    (v, pv) <- ptr_to_fresh_readonly "nonce" (llvm_array 8 (llvm_int 8));
+    (m, pm) <- ptr_to_fresh "buf" (llvm_array n (llvm_int 8));
+
+    crucible_execute_func [ pkey
+                          , pv
+                          , crucible_term {{ 0 : [32] }}
+                          , pm
+                          , crucible_term {{ `n : [32] }}
+                          ];
+
+    crucible_points_to pm (crucible_term {{ Salsa20_encrypt (key, v, m) }});
+    crucible_return (crucible_term {{ 0 : [32] }});
+};
+~~~~
+
+### Verifying Everything
+
+Finally, we can verify all of the functions. Notice the use of compositional
+verification and that path satisfiability checking is enabled for those
+functions with loops not bounded by explicit constants.
```

Comment:
Maybe comment on the fact that we prove the top-level function for several sizes?